### ISSUE TITLE
Am I supposed to use string or symbol keys for skills/dependencies?

### ISSUE THREAD

**Issue by georgemillo** [trailblazer/trailblazer-operation#13]
The title says it all. The docs aren't clear on this, and the behaviour seems somewhat inconsistent. When I pass in a dependency in at run-time, I can use it as a KW arg whether I used a string key or a symbol key:

```ruby
class MyOp < Trailblazer::Operation
  step ->(opts, something:, **) { opts['foo'] = something }
end

MyOp.({}, something: 'bar')['foo']
# => "bar" 
MyOp.({}, 'something' => 'bar')['foo']
# => "bar" 
```

But if I want to access `something` via the `options` hash (rather than via KW arg), I have to think about whether I'm using a string or a symbol:

```ruby
class MyOp < Trailblazer::Operation
  success ->(opts) { opts['string'] = opts['something'] }
  success ->(opts) { opts['symbol'] = opts[:something] }
end

sym_result = MyOp.({}, something: true)
sym_result['string']
# => nil
sym_result['symbol']
# => true

srt_result = MyOp.({}, 'something' => true)
str_result['string']
# => true
str_result['symbol']
# => nil
```

... which I guess makes sense, but it feels rather inconsistent given my first code example above.

Also, if I'm setting dependencies on the class level, symbol keys don't work like I would have expected:

```ruby
class MyOp < Trailblazer::Operation
  self[:symbol_key]  = Object

  step ->(opts, symbol_key:, **) { puts symbol_key }
end

MyOp.()
# => ArgumentError: missing keyword: symbol_key

# Although I can still pass it in at runtime:
MyOp.({}, symbol_key: 123)
# => 123
```

The docs seem to exclusively use string keys for class-level dependencies, so maybe there's a good reason why I shouldn't be using symbol keys in the first place. But why not make something like this possible?

```ruby
class OpWithClassLevelDependency < Trailblazer::Operation
  self['dependency']  = Object

  step ->(opts, dependency:, **) { puts dependency }
end

# Let me use the class dependency by default (right now this crashes):
OpWithClassLevelDependency.()
# => Object
# ... but also let my override the default and use KW args in the method signature:
OpWithClassLevelDependency.({}, dependency: BasicObject)
# BasicObject
```

Should I just use string keys everywhere? If so, why not either a) enforce it so that trying to set a skill with a symbol key raises an error or b) let users set symbol keys if they want, but make it clearer in the docs that this is discouraged?

And in any case, can the "make skills accessible as KW args" magic be expanded so that something like my `OpWithClassLevelDependency` class above will work?

**Comment by apotonick** [comment]
In 2.1, we're introducing "hash with indifferent access" so it's basically up to you. We have corrected that in the new website/docs. Right, @yogeshjain999 ? :ice_cream:

**Comment by yogeshjain999** [comment]
Snap! We've section for that in website, but we need to add the details. Making a note about this 🎵

**Comment by panSarin** [comment]
Since that issue is resolved in 2.1, for now you can check current documentation under https://2019.trailblazer.to/ which new - more complete version will be published during next days. I am clossing issue, but feel free to ask any questions if you have any doubts. Also join our zulipchat under https://trailblazer.zulipchat.com/
